src/NuGet.Core/NuGet.DependencyResolver.Core/ResolverUtility.cs

```diff
@@ -29,8 +28,17 @@ public static class ResolverUtility
         {
             var key = new LibraryRangeCacheKey(libraryRange, framework);
 
-            return cache.GetOrAdd(key, (cacheKey) =>
-                FindLibraryEntryAsync(cacheKey.LibraryRange, framework, runtimeIdentifier, context, cancellationToken));
+            if (!cache.TryGetValue(key, out var graphItem))
+            {
+                graphItem = FindLibraryEntryAsync(key.LibraryRange, framework, runtimeIdentifier, context, cancellationToken);
+
+                if (!cache.TryAdd(key, graphItem))
```

Comment:
```Suggestion
                cache.TryAdd(key, graphItem);
```
Currently we try to add then it'll fail only if item was already in dictionary which is very unlikely, even it's the case  those task still return same result. So could you be able to do one more perf profiling with above suggestion and share the result?

Reply:
I tweaked your suggestion a bit and observed memory allocation reduction by 208 MB. My understanding may be something went wrong when I captured the PerfView PR initially. I followed following steps now.

- Open `OrchardCore` solution in VS.
- Delete complete NuGet cache.
- Start collecting PerfView trace
- Right click on solution explorer -> restore nuget packages
- Analyzed results

1. The existing code in the `dev` branch has a memory allocation of `364 MB`.
https://github.com/NuGet/NuGet.Client/blob/8387514417648d4ed1c5e57a6bd32fef506f0b71/src/NuGet.Core/NuGet.DependencyResolver.Core/ResolverUtility.cs#L30-L33

![image](https://user-images.githubusercontent.com/52756182/127720588-610eca25-f62c-4383-ab6d-920517331678.png)

2. The following code caused an allocation of `173 MB` which is `191 MB` less than current `dev` branch code.

``` cs
            var key = new LibraryRangeCacheKey(libraryRange, framework);

            if (!cache.TryGetValue(key, out var graphItem))
            {
                graphItem = FindLibraryEntryAsync(key.LibraryRange, framework, runtimeIdentifier, context, cancellationToken);

                if (!cache.TryAdd(key, graphItem))
                {
                    graphItem = cache[key];
                }
            }

            return graphItem;
```
![image](https://user-images.githubusercontent.com/52756182/127720742-ad5f5f0a-5bf6-4aee-b8a7-3a4829020913.png)

3. The following code caused an allocation of `156 MB` which is `208` less than current `dev` branch.
```cs
            var key = new LibraryRangeCacheKey(libraryRange, framework);

            if (cache.TryGetValue(key, out var graphItem))
                return graphItem;

            _ = cache.TryAdd(key, FindLibraryEntryAsync(key.LibraryRange, framework, runtimeIdentifier, context, cancellationToken));

            return cache[key];
```
![image](https://user-images.githubusercontent.com/52756182/127720780-91429f38-4382-422d-a09c-e021fe47f5ac.png)

Given that the 3rd option gave best results, I pushed a new commit with these changes.